arangod/Replication2/ReplicatedLog/LogLeader.cpp

```diff
@@ -0,0 +1,945 @@
+}
+
+auto replicated_log::LogLeader::tryHardToClearQueue() noexcept -> void {
+  bool finished = false;
+  auto consecutiveTriesWithoutProgress = 0;
+  do {
+    ++consecutiveTriesWithoutProgress;
+    try {
+      auto leaderDataGuard = acquireMutex();
+      auto& queue = leaderDataGuard->_waitForQueue;
+      // The queue cannot be empty: resign() clears it while under the mutex,
+      // and waitFor also holds the mutex, but refuses to add entries after
+      // the leader resigned.
+      // This means it should never happen in production code. But this
+      // assertion is really annoying in the tests.
+#ifndef ARANGODB_USE_GOOGLE_TESTS
+      TRI_ASSERT(queue.empty());
+#endif
+      if (!queue.empty()) {
+        LOG_CTX("8b8a2", ERR, _logContext)
+            << "Leader destroyed, but queue isn't empty!";
+        for (auto it = queue.begin(); it != queue.end();) {
+          if (!it->second.isFulfilled()) {
+            it->second.setException(basics::Exception(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED,
+                                                      __FILE__, __LINE__));
+          } else {
+            LOG_CTX("002b2", ERR, _logContext)
+                << "Fulfilled promise in replication queue!";
+          }
+          it = queue.erase(it);
+          consecutiveTriesWithoutProgress = 0;
+        }
+      }
+      finished = true;
+    } catch (basics::Exception const& exception) {
+      LOG_CTX("eadb8", ERR, _logContext)
+          << "Caught exception while destroying a log leader: " << exception.message();
+    } catch (std::exception const& exception) {
+      LOG_CTX("35d0b", ERR, _logContext)
+          << "Caught exception while destroying a log leader: " << exception.what();
+    } catch (...) {
+      LOG_CTX("0c972", ERR, _logContext)
+          << "Caught unknown exception while destroying a log leader!";
+    }
+    if (!finished && consecutiveTriesWithoutProgress > 10) {
+      LOG_CTX("d5d25", FATAL, _logContext)
+          << "We keep failing at destroying a log leader instance. Giving up "
+             "now.";
+      FATAL_ERROR_EXIT();
+    }
+  } while (!finished);
+}
+
+auto replicated_log::LogLeader::instantiateFollowers(
+    LoggerContext logContext, std::vector<std::shared_ptr<AbstractFollower>> const& follower,
+    std::shared_ptr<LocalFollower> const& localFollower, TermIndexPair lastEntry)
+    -> std::vector<FollowerInfo> {
+  auto initLastIndex = lastEntry.index.saturatedDecrement();
+
+  std::vector<FollowerInfo> follower_vec;
+  follower_vec.reserve(follower.size() + 1);
+  follower_vec.emplace_back(localFollower, lastEntry, logContext);
+  std::transform(follower.cbegin(), follower.cend(), std::back_inserter(follower_vec),
+                 [&](std::shared_ptr<AbstractFollower> const& impl) -> FollowerInfo {
+                   return FollowerInfo{impl, TermIndexPair{LogTerm{0}, initLastIndex}, logContext};
+                 });
+  return follower_vec;
```

Comment:
Would love `follower` to be renamed to plural `followers` here, as it can be multiple followers.